_posts/2017-09-24-Unit-testing-in-angular.md

```diff
@@ -0,0 +1,504 @@
+A test itself starts with `it`. 
+Besides tests, you can also add other blocks in to a `describe`, like `beforeEach`, `beforeAll`, `afterEach`, `afterAll`... 
+What these blocks do, is quite self-explanatory. 
+Here's an example how it could be used: when testing a class, you’ll want to create an instance of that class for each test, so instead of writing the same code in each test, you could put that code in the `beforeEach` clause. 
+Simply pass `beforeEach` a function (in lambda notation) containing the code you want it to run. 
+
+Within a test itself, the class' public functions can be called and assertions can be made. 
+Assertions are made using `expect`. You can tell it what you expect the result of what you have given it to be with `toBe`, `toEqual`, `toBeTruthy`, `toBeFalsy`, `toBeNull`...
+
+Here's an example: 
+
+{% highlight coffeescript %}  
+    describe('NAME_OF_YOUR_CLASS', () => {
+        let component;
+
+        beforeEach(() => {
+            //initialize
+            component = new AppComponent();
+        });
+        
+        //Actual tests
+        it('should have a car selected', () => {
+            //assertions
+            expect(component.carSelected).toBeThruthy();
+        });
+    
+        it('should find my favorite car brand', () => {
+            //assertions
+            const carBrand = component.getFavoriteCarBrand();
+            expect(carBrand.name).toEqual('Mazda');
+        }); 
+    }); 
+{% endhighlight %}  
+
+As you can see, you can pass some text as an argument in the `describe` call. 
+This is usually the name of the class you're testing and it'll be shown when running the tests. 
+For the tests themselves, you can also pass some text which will be shown. 
+These are mainly used for you to be able to identify the tests if they fail. 
+The text should describe what's being tested, like "It should get the brand of the car", could be written as `it('should get the brand of the car', () => ...`.
+
+## Writing the actual tests
+There are two ways to write unit tests for an Angular app. 
+Either you use the Angular TestBed or you simply call the constructor of the class directly. 
+
+## 1. TestBed
+TestBed is something pretty cool Angular came up with in order to test your components. 
+It allows to create an instance of a component and injects all dependencies listed in the constructor of that component's class. 
+However, you do need to list all dependencies in the TestBed. 
+It kind of looks like a module definition:
+
+{% highlight coffeescript %}  
+    describe('AppComponent', () => {
+        let component: AppComponent;
+        let fixture: ComponentFixture<AppComponent>;
+
+        beforeEach(async(() => {
+            TestBed.configureTestingModule({
+                declarations: [AppComponent],
+                providers: [CarBrandService],
+                imports: [CommonLogicModule]
+            })
+            .compileComponents();
```

Comment:
Note: WebPack developers need not call compileComponents because it inlines templates and css as part of the automated build process that precedes running the test.